pkg/cloud/services/autoscaling/autoscalinggroup_test.go

```diff
@@ -220,3 +317,722 @@ func TestService_SDKToAutoScalingGroup(t *testing.T) {
+							Value:             aws.String("owned"),
+						},
+						{
+							Key:               aws.String("sigs.k8s.io/cluster-api-provider-aws/cluster/test"),
+							PropagateAtLaunch: aws.Bool(false),
+							ResourceId:        aws.String("create-asg-success"),
+							ResourceType:      aws.String("auto-scaling-group"),
+							Value:             aws.String("owned"),
+						},
+						{
+							Key:               aws.String("sigs.k8s.io/cluster-api-provider-aws/role"),
+							PropagateAtLaunch: aws.Bool(false),
+							ResourceId:        aws.String("create-asg-success"),
+							ResourceType:      aws.String("auto-scaling-group"),
+							Value:             aws.String("node"),
+						},
+						{
+							Key:               aws.String("Name"),
+							PropagateAtLaunch: aws.Bool(false),
+							ResourceId:        aws.String("create-asg-success"),
+							ResourceType:      aws.String("auto-scaling-group"),
+							Value:             aws.String("create-asg-success"),
+						},
+					},
+					VPCZoneIdentifier: aws.String("subnet1"),
+				}
+
+				m.CreateAutoScalingGroup(gomock.AssignableToTypeOf(&autoscaling.CreateAutoScalingGroupInput{})).Do(
+					func(actual *autoscaling.CreateAutoScalingGroupInput) (*autoscaling.CreateAutoScalingGroupOutput, error) {
+						sortTagsByKey := func(tags []*autoscaling.Tag) {
+							sort.Slice(tags, func(i, j int) bool {
+								return *(tags[i].Key) < *(tags[j].Key)
+							})
+						}
+						// sorting tags to avoid failure due to different ordering of tags
+						sortTagsByKey(actual.Tags)
+						sortTagsByKey(expected.Tags)
+						if !reflect.DeepEqual(expected, actual) {
+							t.Fatalf("Actual CreateAutoScalingGroupInput did not match expected, Actual : %v, Expected: %v", actual, expected)
+						}
+						return &autoscaling.CreateAutoScalingGroupOutput{}, nil
+					})
+			},
+			check: func(autoscalinggroup *expinfrav1.AutoScalingGroup, err error) {
+				if err != nil {
+					t.Fatalf("did not expect error: %v", err)
+				}
+				if autoscalinggroup != nil {
+					t.Fatalf("expected asg to be nil but got: %+v", autoscalinggroup)
+				}
+			},
+		},
+		{
+			name:            "should return error if create ASG fails",
+			machinePoolName: "create-asg-fail",
+			expect: func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder) {
+				s.Name().Times(2).Return("test")
+				s.Info(gomock.Eq("Running instance"))
+				m.CreateAutoScalingGroup(gomock.AssignableToTypeOf(&autoscaling.CreateAutoScalingGroupInput{})).Return(nil, awserrors.NewFailedDependency("dependency failure"))
+				s.Error(gomock.Any(), gomock.Eq("unable to create AutoScalingGroup"))
+			},
+			check: func(autoscalinggroup *expinfrav1.AutoScalingGroup, err error) {
+				if err == nil {
+					t.Fatalf("expected error to occur")
+				}
+				if autoscalinggroup != nil {
+					t.Fatalf("expected asg to be nil but got: %+v", autoscalinggroup)
+				}
+			},
+		},
```

Comment:
Can we also cover the following scenarios:
- should fail in getting subnets for ASG
- should fail if AWSMachinePool has no LaunchTemplateID